### PR TITLE
support loading pipeline from transformer style (flat) repo 

### PR DESCRIPTION
**Pull request by yiyixuxu**
With this PR, we want to support loading a pipeline from [`google/diffusiongemma-26B-A4B-it`](https://huggingface.co/google/diffusiongemma-26B-A4B-it), and also future model releases in the same style :
* model released in transformers with the flat structure, i.e., model and processor are hosted at the root folder directly
* pipeline + scheduler in diffusers, and would be able to add a scheduler folder and `model_index.json` file into the same repo

Currently, we have some legacy support for loading components from the root, but very limited (e.g., it does not work if the repo also has a subfolder, or if your component comes with files other than the regular weights, etc.) . Since all our releases use subfolder structure, this was sufficient as it is (I'm actually not sure why the legacy support was there at all; I assume a very early model released that way, and we kept it for backward compatibility).

This PR extends that a bit further to support joint transformer + diffusers release (we will likely see more)
*  make it work with a mixed structure (some components at the root, some in subfolders). Here, `model` and `processor` live in the root, `scheduler` has its own folder)
* always allow a fixed list of known transformers auxiliary filenames at the root (`TRANSFORMERS_COMPONENT_AUX_FILES`: tokenizer/processor/chat-template/generation configs etc), currently, we only allow the list of files DiffusionGemma requires, but we can extend as we support more pipelines

you can test with this
```py
import torch
from diffusers import DiffusionGemmaPipeline

# model_id = "google/diffusiongemma-26B-A4B-it" 
model_id = "dg845/diffusiongemma-test"

pipe = DiffusionGemmaPipeline.from_pretrained(
    model_id,
    torch_dtype=torch.bfloat16,
    # revision="refs/pr/20",   # Omit if merged to main
)
pipe.to("cuda")

output = pipe(
    prompt="Why is the sky blue?",
    gen_length=256,
    num_inference_steps=48,
    cache_implementation="static",
    generator=torch.Generator("cuda").manual_seed(42),
)
print(output.texts[0])
```

### CODE REVIEW

File: src/diffusers/pipelines/pipeline_utils.py [hunk 1]
```diff
@@ -1751,6 +1752,11 @@ def download(cls, pretrained_model_name, **kwargs) -> str | os.PathLike:
                 p for p in allow_patterns if not (len(p.split("/")) == 2 and p.split("/")[0] in passed_components)
             ]
 
+            # Files explicitly allow-listed by the repo author via `_allow_files` in `model_index.json` are
+            # added to the download set. This supports repos that keep a component's config/tokenizer files at
+            # the root (instead of in its own subfolder), where the folder-based allow patterns would miss them.
+            allow_patterns += allow_filenames
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/pipeline_utils.py]: need to work with field like this https://huggingface.co/YiYiXu/diffusiongemma-test/blob/main/model_index.json#L4

Comment (yiyixuxu) [reply src/diffusers/pipelines/pipeline_utils.py]: alternative is we could just add these processor-related files into the default `allow_patterns` here (without the model author having to specify in `model_index.json`), I leaned toward the design in this PR mainly because:

(1) it's more practical and more remote friendly - for a repo hosted this way with slightly different files, the model author can just update the `model_index.json` instead of PR into diffusers 
(2) more fundamentally, I don't want to broadly support loading components from the roots - it is just too ambiguous for diffusers because we have multiple components and we often co-host our checkpoints in the same repo as the research repo's original checkpoint and their checkpoint is hosted at the root.  We already have some implicit root-loading support today, but I don't think we need to expand it as default behavior. `_allow_files` keeps it explicit and opt-in

But I'm open to different opinions!

File: tests/pipelines/test_pipeline_utils.py [hunk 2]
```diff
@@ -218,6 +218,34 @@ def test_diffusers_is_compatible_no_components_only_variants(self):
...
+        ]
+        self.assertTrue(is_safetensors_compatible(filenames))
+
+    def test_diffusers_is_compatible_no_components_safetensors_only_variants(self):
+        filenames = [
+            "diffusion_pytorch_model.fp16.safetensors",
+        ]
+        self.assertTrue(is_safetensors_compatible(filenames, variant="fp16"))
+
+    def test_diffusers_is_compatible_weightless_subfolder(self):
```
Comment (yiyixuxu) [inline tests/pipelines/test_pipeline_utils.py]: only this test will have a different result from `main`
Currently, in `main` this will marked as not `is_safetensors_compatible`

File: src/diffusers/pipelines/pipeline_loading_utils.py [hunk 3]
```diff
@@ -136,6 +136,8 @@ def is_safetensors_compatible(filenames, passed_components=None, folder_names=No
     )
 
     passed_components = passed_components or []
+    # only weight files matter for safetensors compatibility
+    filenames = filter_model_files(filenames)
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/pipeline_loading_utils.py:153]: i think it's a bug here but let me know cc @DN6

**Review by dg845 — APPROVED** [review]
Thanks for the PR!

File: src/diffusers/pipelines/pipeline_utils.py (hunk 1 above)
Comment (sayakpaul) [reply src/diffusers/pipelines/pipeline_utils.py]: The current one is better as it gives the idea that model authors need to be explicit about this kind of structure. Loading it implicitly will lead to confusion.

File: tests/pipelines/test_pipeline_utils.py [hunk 4]
```diff
@@ -218,6 +218,34 @@ def test_diffusers_is_compatible_no_components_only_variants(self):
...
+        # transformers-style flat layout: safetensors weights at the root + a weight-less subfolder (scheduler/)
+        filenames = [
+            "diffusion_pytorch_model.safetensors",
```
Comment (sayakpaul) [inline tests/pipelines/test_pipeline_utils.py]: Should we follow the transformers-style weight naming convention here? (`model-00001-of-00011.safetensors`, and so on)?

**Review by sayakpaul — APPROVED** [review]

File: tests/pipelines/test_pipeline_utils.py (hunk 2 above)
Comment (DN6) [reply tests/pipelines/test_pipeline_utils.py]: Oh so main is wrong here? This test should evaluate to `compatible == True`.

File: src/diffusers/pipelines/pipeline_loading_utils.py (hunk 3 above)
Comment (DN6) [reply src/diffusers/pipelines/pipeline_loading_utils.py:153]: LGTM 👍🏽

File: src/diffusers/pipelines/pipeline_utils.py (hunk 1 above)
Comment (DN6) [reply src/diffusers/pipelines/pipeline_utils.py]: How does one define the allow list when saving the model? My one concern here is that this introduces a parameter that allows a repo owner to change which files are downloaded from the hub repo onto a users machine without the end user explicitly opting in for it. e.g. I can add `_allow_files=['weird.pt']`  and this would end up on a users machine. 

Allowing root level components is probably not ideal, but in this case we'd just be expanding to `transformers` components and we would be unlikely to add more file types. IMO this tradeoff is okay because with `_allow_file` we would be enabling downloading potentially any file type without explicit opt in.

Comment (yiyixuxu) [reply src/diffusers/pipelines/pipeline_utils.py]: ohh that's really good point!

**Review by DN6 — APPROVED** [review]
Thanks 👍🏽
